clients/src/main/java/org/apache/kafka/common/utils/Utils.java

```diff
@@ -1098,6 +1099,29 @@ public static void writeFully(FileChannel channel, ByteBuffer sourceBuffer) thro
             channel.write(sourceBuffer);
     }
 
+    /**
+     * Trying to write data in source buffer to a {@link TransferableChannel}, we may need to call this method multiple
+     * times since this method doesn't ensure data in source buffer can be fully written to dest channel.
+     *
+     * @param destChannel The dest channel
+     * @param position From which the source buffer will be written
+     * @param length The max size of bytes can be written
+     * @param sourceBuffer The source buffer
+     *
+     * @return The length of the actual written data
+     * @throws IOException If an I/O error occurs
+     */
+    public static long tryWriteTo(TransferableChannel destChannel,
+                                  int position,
+                                  int length,
+                                  ByteBuffer sourceBuffer) throws IOException {
```

Comment:
How about removing the IO from this method with a signature like:
```java
public static ByteBuffer relativeSlice(ByteBuffer buffer, int position, int length);
```
The caller of this method can then `long written = destChannel.write(Utils.relativeSlice(buffer, position, length))`.